### PR TITLE
General improvements

### PR DESCRIPTION
**Pull request by lildude**
## Added

- Devcontainer configuration for easier local dev which can also be used for Codespaces.

## Changed

- Save the JSON pretty-printed when `DEBUG_LOGGING` is enabled for easier debugging
- Changed the default input, output and cache directory path to `/var/tmp/e2g-(input|output|cache)` as not everyone has write access to `/var` when running locally (or in the devcontainer/Codespace where we deliberately don't run as root)
- Manual runs save the file to `/var/tmp/e2g-output` or the specified output dir instead of repo root so easily found and ignored by git
- YahooFinanceService will retry with GBp if a symbol/ISIN isn't found with GBP as the UK has this annoying habit of mixing GBP and GBp but platforms like Freetrade don't differentiate.
- Removed a load of unnecessary white space - unplanned, but my VSCode config does this for me when I touch files.
- Moved samples to their own directory for a cleaner layout.
- Only run the docker builds in CI if the secrets are set.
- Reordered converters and watcher headers alphabetically as it makes it easier to maintain.

 
## Fixes

- Removed duplicate `.json` in several messages.

## Checklist

- [x] Added relevant changes to README (if applicable)
- [x] Added relevant test(s)
- [x] Updated GitVersion file

### CODE REVIEW

**Comment by lildude** [comment]
🤔 I'm not sure why this is suddenly failing. It looks like a permissions issue, however you're already setting the `pull-requests: write` permission in this workflow 🤷.

**Comment by dickwolff** [comment]
> 🤔 I'm not sure why this is suddenly failing. It looks like a permissions issue, however you're already setting the `pull-requests: write` permission in this workflow 🤷.

I noticed this some time ago, so I created the separate issue #56 to look into this. At first glance, it looks like the settings are correct but somehow it doesn't work.

File: src/manual.ts [hunk 1]
```diff
@@ -11,12 +12,18 @@ else {
 
     // Define import file path.
     const inputFile = process.env.INPUT_FILE;
+    const outputFolder = process.env.E2G_OUTPUT_FOLDER || "tmp/e2g-output";
```
Comment (dickwolff) [inline src/manual.ts:15]: I would make this just `e2g-output` (without the `tmp` for the manual runner. No need to put it two folders deep. Probably also add that folder to the gitignore

File: src/converters/freetradeConverter.test.ts [hunk 2]
```diff
@@ -6,6 +6,14 @@ describe("freetradeConverter", () => {
 
   beforeEach(() => {
     jest.spyOn(console, "log").mockImplementation(jest.fn());
+    // Mock YahooFinanceService - we don't want to make real requests and don't care about the result.
```
Comment (dickwolff) [inline src/converters/freetradeConverter.test.ts:9]: I certainly do see the appeal of this, as it does shorten the test time significantly.  I also agree with your statement that the YahooFinance converter should get it's own test file. I haven't gotten to that just yet 😄

However I feel this approach might be a little simplistic regarding testing the converters. I think the coverage of this approach will drop below the current %. The "beauty" (or not, depending on how you look at it 😉) of the current approach is that it magically covers (almost) all branches in the converter. I feel a better (but for sure more time consuming approach) would be to register all the current calls that are done to Yahoo Finance, register all the responses and put those in a (generic) mock file that can be used in all of the converter tests. This way you have the benefit of mocking (and not calling Yahoo Finance, thus making it faster) and also the guaranteed repeatability of the test.

File: src/converters/degiroConverterV2.test.ts [hunk 3]
```diff
@@ -5,13 +5,21 @@ import { GhostfolioExport } from "../models/ghostfolioExport";
 describe("degiroConverterV2", () => {
 
   beforeEach(() => {
-   // jest.spyOn(console, "log").mockImplementation(jest.fn());
+    // jest.spyOn(console, "log").mockImplementation(jest.fn());
+    // Mock YahooFinanceService - we don't want to make real requests and don't care about the result.
+    jest.spyOn(YahooFinanceService.prototype, "getSecurity").mockResolvedValue({
```
Comment (lildude) [inline src/converters/degiroConverterV2.test.ts:10]: Testing that comments are working using the "Add single comment" button.

File: src/converters/degiroConverterV2.test.ts [hunk 4]
```diff
@@ -5,13 +5,21 @@ import { GhostfolioExport } from "../models/ghostfolioExport";
...
+      symbol: "AAPL",
```
Comment (lildude) [inline src/converters/degiroConverterV2.test.ts:11]: And another using the "Start a review" button and then reviewing using the "Review changes" button at the top right of the code tab.

**Review by lildude — COMMENTED** [review]
Testing

File: src/converters/degiroConverterV2.test.ts (hunk 3 above)
Comment (dickwolff) [reply src/converters/degiroConverterV2.test.ts:10]: And now?

File: src/converters/freetradeConverter.test.ts (hunk 2 above)
Comment (dickwolff) [reply src/converters/freetradeConverter.test.ts]: Added #58 for this purpose. When I have some time I will make a start with this.

File: src/converters/freetradeConverter.ts
Comment (dickwolff) [inline src/converters/freetradeConverter.ts:1]: Apparently I can't comment on the exact lines, so I'll put the comment on the file level. I had this comment on your draft PR but somehow forgot to copy it over.

If I'm right, the lines from 123-130 can be put in with line 32. This way the type processing takes place beforehand.

**Comment by dickwolff** [comment]
> 🤔 I'm not sure why this is suddenly failing. It looks like a permissions issue, however you're already setting the `pull-requests: write` permission in this workflow 🤷.

As a test for the test runner, can you check the Actions > General settings of your fork and see if they are the same as mine?

image

File: .github/workflows/docker.yml [hunk 5]
```diff
@@ -17,36 +17,41 @@ on:
   workflow_dispatch: # Also allow manual triggering
 
 jobs:
-  docker:
+  docker:    
+    env:
+      dockerhub_username: ${{ secrets.DOCKERHUB_USERNAME }}
+      dockerhub_token: ${{ secrets.DOCKERHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (lildude) [reply .github/workflows/docker.yml:23]: Doh! 🤦 Thanks.

File: src/converters/freetradeConverter.ts
Comment (lildude) [reply src/converters/freetradeConverter.ts:1]: For the dividend, yes definitely.

I'm not sure about the buy/sell. How do I get the value from another column for each line?

Comment (lildude) [reply src/converters/freetradeConverter.ts:1]: I've just noticed the dividend bit isn't needed so I've removed it.

**Comment by dickwolff** [comment]
Merged and released under 0.10.0.

Thanks for the contribution!!

**Pull request merged**